### PR TITLE
Add TSV input funcionality and extra goodies.

### PR DESCRIPTION
**Pull request by jfy133**
# nf-core/eager pull request

This PR mainly includes everything related to our new input system, which allows either direct paths to FASTQs/BAMS, or alternatively, with a TSV 'map' which contains all the relevent per-library metadata.

Along with this is updated documentation, lots of minor bug fixes including emailing, profile loading etc.

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If necessary, also make a PR on the [nf-core/eager branch on the nf-core/test-datasets repo](https://github.com/nf-core/test-datasets/pull/new/nf-core/eager)
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --paired_end`).
- [x] Make sure your code lints ([`nf-core lint .`](https://nf-co.re/tools)).
- [x] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [x] `README.md` is updated

**Learn more about contributing:** [CONTRIBUTING.md](https://github.com/nf-core/eager/tree/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -1270,244 +1404,309 @@ process samtools_filter {
...
-    if (params.single_end) {
+    if ( seqtype == 'SE' ) {
         out_fwd = bam.baseName+'.stripped.fq.gz'
         """
         samtools index $bam
-        extract_map_reads.py $bam ${fq[0]} -m ${params.strip_mode} -of $out_fwd -p ${task.cpus}
+        extract_map_reads.py $bam ${r1} -m ${params.strip_mode} -of $out_fwd -p ${task.cpus}
         """
     } else {
         out_fwd = bam.baseName+'.stripped.fwd.fq.gz'
```
Comment (maxibor) [inline main.nf:1483]: Change to `libraryid`

File: main.nf [hunk 2]
```diff
@@ -1270,244 +1404,309 @@ process samtools_filter {
...
         out_rev = bam.baseName+'.stripped.rev.fq.gz'
```
Comment (maxibor) [inline main.nf:1484]: same

File: main.nf [hunk 3]
```diff
@@ -1270,244 +1404,309 @@ process samtools_filter {
...
 
     output:
-    tuple val(prefix), file("*stats") into ch_bam_filtered_flagstat_for_multiqc, ch_bam_filtered_flagstat_for_endorspy
+    tuple samplename, libraryid, lane, seqtype, organism, strandedness, udg, file("*.stats") into ch_bam_filtered_flagstat_for_multiqc, ch_bam_filtered_flagstat_for_endorspy
 
     script:
-    prefix = "$bam" - ~/(\.bam.filtered.bam)?$/
+    prefix = libraryid
     """
     samtools flagstat $bam > ${prefix}_postfilterflagstat.stats
```
Comment (maxibor) [inline main.nf:1512]: Can directly use `libraryid` instead of `prefix`

File: main.nf [hunk 4]
```diff
@@ -1270,244 +1404,309 @@ process samtools_filter {
...
+
+      [samplename, libraryid, lane, seqtype, organism, strandedness, udg, stats, poststats ] }
     .set{ ch_allflagstats_for_endorspy }
 }
 
+// Endogenous DNA calculator to say how much of a library contained 'on-target' DNA
+
 process endorSpy {
     label 'sc_tiny'
     tag "$prefix"
```
Comment (maxibor) [inline main.nf:1543]: `tag "$libraryid"`

File: main.nf [hunk 5]
```diff
@@ -1270,244 +1404,309 @@ process samtools_filter {
...
-    tuple val(name), file(stats), file(poststats) from ch_allflagstats_for_endorspy
+    tuple samplename, libraryid, lane, seqtype, organism, strandedness, udg, file(stats), file(poststats) from ch_allflagstats_for_endorspy
 
     output:
-    file "*.json" into ch_endorspy_for_multiqc
+    tuple samplename, libraryid, lane, seqtype, organism, strandedness, udg, file("*.json") into ch_endorspy_for_multiqc
 
     script:
-    prefix = "${name}"
+    prefix = libraryid
```
Comment (maxibor) [inline main.nf:1553]: remove

**Comment by maxibor** [comment]
General comment: there are many `prefix` that you can completely replace by `libraryid` in many processes.

File: main.nf [hunk 6]
```diff
@@ -278,189 +287,163 @@ if( params.bwa_index && (params.mapper == 'bwaaln' | params.mapper == 'bwamem'))
...
-        exit 1, "--strip_mode can only be set to strip or replace!"
-    }
-
-    if (params.bam && !params.run_convertbam) {
-        exit 1, "--strip_input_fastq can only be used on FASTQ, but you gave BAM input and didn't specify --run_convertbam!"
+        exit 1, "[nf-core/eager] error: --strip_mode can only be set to strip or replace!"
     }
 }
 
 // Mapper sanity checking
```
[the remark this answers is not in the document]
Comment (jfy133) [reply main.nf:317]: 397:
```
// Validate that skip_collapse is only set to True for paired_end reads!
if (!has_extension(params.input, "tsv") && params.skip_collapse  && params.single_end){
    exit 1, "[nf-core/eager] error: --skip_collapse can only be set for paired_end samples!"
}
```

280:
```
// Check if genome exists in the config file. params.genomes is from igenomes.conf, params.genome specified by user
if (params.genomes && params.genome && !params.genomes.containsKey(params.genome)) {
    exit 1, "[nf-core/eager] error: the provided genome '${params.genome}' is not available in the iGenomes file. Currently the available genomes are ${params.genomes.keySet().join(", ")}"
}

```

File: main.nf [hunk 7]
```diff
@@ -278,189 +287,163 @@ if( params.bwa_index && (params.mapper == 'bwaaln' | params.mapper == 'bwamem'))
...
-    exit 1, "Unknown MALT mode specified. Options: 'BlastN', 'BlastP', 'BlastX'. You gave '${params.malt_mode}'!"
+    exit 1, "[nf-core/eager] error: unknown MALT mode specified. Options: 'BlastN', 'BlastP', 'BlastX'. You gave '${params.malt_mode}'!"
   }
 
   if (params.metagenomic_tool == 'malt' && params.malt_alignment_mode != 'Local' && params.malt_alignment_mode != 'SemiGlobal') {
-    exit 1, "Unknown MALT alignment mode specified. Options: 'Local', 'SemiGlobal'. You gave '${params.malt_alignment_mode}'!"
+    exit 1, "[nf-core/eager] error: unknown MALT alignment mode specified. Options: 'Local', 'SemiGlobal'. You gave '${params.malt_alignment_mode}'!"
   }
 
   if (params.metagenomic_tool == 'malt' && params.malt_min_support_mode == 'percent' && params.metagenomic_min_support_reads != 1) {
```
[the remark this answers is not in the document]
Comment (jfy133) [reply main.nf:419]: This is a check that someone has tried ot change the min_support_reads (away from our default of 1), but specified they want to use `percent` method as their min cut off. It's not a input data issue.

I'm assume you're meaning this check?
```
  if (params.metagenomic_tool == 'malt' && params.malt_min_support_mode == 'percent' && params.metagenomic_min_support_reads != 1) {
    exit 1, "[nf-core/eager] error: incompatible MALT min support configuration. Percent can only be used with --malt_min_support_percent. You modified --metagenomic_min_support_reads!"
  }
```

File: main.nf [hunk 8]
```diff
@@ -278,189 +287,163 @@ if( params.bwa_index && (params.mapper == 'bwaaln' | params.mapper == 'bwamem'))
...
-    exit 1, "Incompatible MALT min support configuration. Percent can only be used with --malt_min_support_percent. You modified --metagenomic_min_support_reads!"
+    exit 1, "[nf-core/eager] error: incompatible MALT min support configuration. Percent can only be used with --malt_min_support_percent. You modified --metagenomic_min_support_reads!"
   }
 
   if (params.metagenomic_tool == 'malt' && params.malt_min_support_mode == 'reads' && params.malt_min_support_percent != 0.01) {
```
[the remark this answers is not in the document]
Comment (jfy133) [reply main.nf:423]: Reverse of the above.

File: main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxibor) [reply main.nf:317]: The line number don't match anymore... :(

[the remark this answers is not in the document]
Comment (jfy133) [reply main.nf:317]: Found it!

**Pull request merged**